master/internal/db/postgres_trial.go

```diff
@@ -279,6 +554,120 @@ func (db *PgDB) AddValidationMetrics(
 	return db.addTrialMetrics(ctx, m, true)
 }
 
+const (
+	// InfPostgresString how we store infinity in JSONB in postgres.
+	InfPostgresString = "Infinity"
+	// NegInfPostgresString how we store -infinity in JSONB in postgres.
+	NegInfPostgresString = "-Infinity"
+	// NaNPostgresString how we store NaN in JSONB in postgres.
+	NaNPostgresString = "NaN"
+)
+
+func stringToSpecialFloats(s string) (float64, bool) {
+	switch s {
+	case NaNPostgresString:
+		return math.NaN(), true
+	case InfPostgresString:
+		return math.Inf(1), true
+	case NegInfPostgresString:
+		return math.Inf(-1), true
+	default:
+		return 0.0, false
+	}
+}
+
+func replaceSpecialFloatsWithString(v any) any {
+	if f, ok := v.(float64); ok {
+		switch {
+		case math.IsNaN(f):
+			return NaNPostgresString
+		case math.IsInf(f, 1.0):
+			return InfPostgresString
+		case math.IsInf(f, -1.0):
+			return NegInfPostgresString
+		}
+	}
+	return v
+}
+
+func calculateNewSummaryMetrics(
+	summaryMetrics model.JSONObj, metrics *structpb.Struct,
+) model.JSONObj {
+	// Calculate numeric metrics.
+	for metricName, metric := range metrics.Fields {
+		// Add an empty map for any non numeric metrics.
+		metricValue, providedNumeric := metric.AsInterface().(float64)
+		if !providedNumeric {
+			if s, isString := metric.AsInterface().(string); isString {
+				if f, ok := stringToSpecialFloats(s); ok {
+					metricValue = f
+					providedNumeric = true
+				}
+			}
+
+			if !providedNumeric {
+				summaryMetrics[metricName] = map[string]any{}
+				continue
+			}
+		}
+
+		// If we haven't seen the metric before just add it.
+		if summaryMetric, ok := summaryMetrics[metricName].(map[string]any); !ok {
+			summaryMetrics[metricName] = map[string]any{
+				"max":   replaceSpecialFloatsWithString(metricValue),
+				"min":   replaceSpecialFloatsWithString(metricValue),
+				"sum":   replaceSpecialFloatsWithString(metricValue),
+				"count": 1,
+			}
+		} else {
+			// Check if the metric had a non numeric value in the past.
+			notNumeric := false
+			for _, agg := range []string{"min", "max", "sum", "count"} {
+				switch v := summaryMetric[agg].(type) {
+				case float64:
+				case string:
+					f, ok := stringToSpecialFloats(v)
+					if !ok {
+						notNumeric = true
+						continue
+					}
+					summaryMetric[agg] = f
+				default:
+					notNumeric = true
+					break
+				}
+			}
+			if notNumeric {
+				summaryMetrics[metricName] = map[string]any{}
+				continue
+			}
+
+			summaryMetric["min"] = replaceSpecialFloatsWithString(
+				math.Min(summaryMetric["min"].(float64), metricValue))
+			summaryMetric["max"] = replaceSpecialFloatsWithString(
+				math.Max(summaryMetric["max"].(float64), metricValue))
+			summaryMetric["sum"] = replaceSpecialFloatsWithString(
+				summaryMetric["sum"].(float64) + metricValue)
+			// Go parsing odditity treats JSON whole numbers as floats.
```

Comment:
it's [javascript fault](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number): there're no integer values, only floating point numbers